### PR TITLE
Add support for XDP v6

### PR DESCRIPTION
**Pull request by sridhartigera**
## Description

This PR has the following changes
1. Fix IPV6 hostendpoints
2. Support XDP V6 and v6 donottrack policy

## Todos

- [x] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
ebpf: XDP v6 requires Linux kernel 5.18+ (Ubuntu >=22.04)
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/fv/donottrack_test.go [hunk 1]
```diff
@@ -100,13 +111,25 @@ var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ do-not-track policy tests;
 	expectFullConnectivity := func() {
 		cc.ResetExpectations()
 		cc.ExpectSome(tc.Felixes[0], hostW[1].Port(8055))
+		cc.ExpectSome(tc.Felixes[0], hostW[1].Port(8055))
```
Comment (tomastigera) [inline felix/fv/donottrack_test.go:114]: duplicate

File: felix/fv/donottrack_test.go [hunk 2]
```diff
@@ -161,8 +189,21 @@ var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ do-not-track policy tests;
...
+				cc.Expect(None, tc.Felixes[0], hostW[1].Port(8055), ExpectWithIPVersion(6))
+				cc.Expect(None, tc.Felixes[1], hostW[0].Port(8055), ExpectWithIPVersion(6))
+				cc.Expect(Some, tc.Felixes[0], hostW[1].Port(2379), ExpectWithIPVersion(6))
+				cc.Expect(Some, tc.Felixes[1], hostW[0].Port(2379), ExpectWithIPVersion(6))
+				// Port 22 is inbound-only so it'll be blocked by the (lack of egress policy).
+				cc.Expect(None, tc.Felixes[0], hostW[1].Port(22), ExpectWithIPVersion(6))
+				cc.Expect(None, tc.Felixes[1], hostW[0].Port(22), ExpectWithIPVersion(6))
+				// But external client should still be able to access it...
+				cc.Expect(Some, externalClient, hostW[1].Port(22), ExpectWithIPVersion(6))
+				cc.Expect(Some, externalClient, hostW[0].Port(22), ExpectWithIPVersion(6))
```
Comment (tomastigera) [inline felix/fv/donottrack_test.go:203]: Could we put it in a function and use that function to setup the same expectations for either version and just pass in the extra `ExpectWithIPVersion` option? And in other similar places?

File: felix/fv/donottrack_test.go [hunk 3]
```diff
@@ -221,10 +262,9 @@ var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ do-not-track policy tests;
 			Expect(err).NotTo(HaveOccurred())
 
 			expectFullConnectivity()
-
 			if BPFMode() {
 				By("Having a Linux IP set for the egress policy")
-				Expect(tc.Felixes[0].IPSetNames()).To(ConsistOf(utils.IPSetNameForSelector(4, host1Selector)))
+				Expect(tc.Felixes[0].IPSetNames()).To(ContainElements(utils.IPSetNameForSelector(4, host1Selector), utils.IPSetNameForSelector(6, host1Selector)))
```
Comment (tomastigera) [inline felix/fv/donottrack_test.go:267]:
```suggestion
				Expect(tc.Felixes[0].IPSetNames()).To(ContainElements(
				        utils.IPSetNameForSelector(4, host1Selector),
				         utils.IPSetNameForSelector(6, host1Selector),
				  ))
```

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 4]
```diff
@@ -2381,12 +2385,7 @@ func (d *bpfEndpointManagerDataplane) applyPolicyToDataIface(
 
 	go func() {
 		defer parallelWG.Done()
-		// XDP IPv6 is not supported yet.
-		if d.ipFamily == proto.IPVersion_IPV4 {
-			xdpErr = d.attachXDPProgram(xdpAP, ep, state)
-		} else {
-			xdpAP = nil
-		}
+		xdpAP, xdpErr = d.attachXDPProgram(&xdpAttachPoint, ep, state)
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:2389]: Will have 2 conditions in which err is returned. 
1. If we failed to load vX xdp programs in which case, ap returned should be nil.
2. If we failed to load policy, in which case we need to return the AP with proper layout. 

Hence I feel returning ap, err will make life easier.

File: felix/bpf/ut/bpf_prog_test.go [hunk 5]
```diff
@@ -361,7 +373,12 @@ func setupAndRun(logger testLogger, loglevel, section string, rules *polprog.Rul
 	if topts.objname != "" {
 		obj = topts.objname
 	} else if topts.ipv6 {
-		obj += "_v6"
+		if topts.xdp {
+			obj += "_co-re_v6"
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/bpf/ut/bpf_prog_test.go:377]: Need to use _co-re as v6 requires xdp_load_bytes helper.

**Review by tomastigera — APPROVED** [review]

**Pull request merged**